# IS 362 – Week 5 Assignment

In this notebook we will attempt to answer these questions:

1. What is the northernmost airport in the United States?
2. What is the easternmost airport in the United States?
3. On February 12th, 2013, which New York area airport had the windiest weather?

First, we will import the necessary modules:

In [1]:
import pandas
import matplotlib.pyplot as plt
import csv
from IPython.display import Image, HTML

%matplotlib inline

We will be working with two dataframes:

1. airports
2. weather

In [2]:
airports = pandas.read_csv('https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/airports.csv')

In [3]:
airports.head()

faa                           name        lat        lon   alt   tz dst  \
0  04G              Lansdowne Airport  41.130472 -80.619583  1044 -5.0   A   
1  06A  Moton Field Municipal Airport  32.460572 -85.680028   264 -6.0   A   
2  06C            Schaumburg Regional  41.989341 -88.101243   801 -6.0   A   
3  06N                Randall Airport  41.431912 -74.391561   523 -5.0   A   
4  09J          Jekyll Island Airport  31.074472 -81.427778    11 -5.0   A   

              tzone  
0  America/New_York  
1   America/Chicago  
2   America/Chicago  
3  America/New_York  
4  America/New_York

In [4]:
weather = pandas.read_csv('https://raw.githubusercontent.com/hadley/nycflights13/master/data-raw/weather.csv')

In [5]:
weather.head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     1  39.02  26.06  59.37     270.0    10.35702   
1    EWR  2013      1    1     2  39.02  26.96  61.63     250.0     8.05546   
2    EWR  2013      1    1     3  39.02  28.04  64.43     240.0    11.50780   
3    EWR  2013      1    1     4  39.92  28.04  62.21     250.0    12.65858   
4    EWR  2013      1    1     5  39.02  28.04  64.43     260.0    12.65858   

   wind_gust  precip  pressure  visib             time_hour  
0        NaN     0.0    1012.0   10.0  2013-01-01T06:00:00Z  
1        NaN     0.0    1012.3   10.0  2013-01-01T07:00:00Z  
2        NaN     0.0    1012.5   10.0  2013-01-01T08:00:00Z  
3        NaN     0.0    1012.2   10.0  2013-01-01T09:00:00Z  
4        NaN     0.0    1011.9   10.0  2013-01-01T10:00:00Z

## **1. What is the northernmost airport in the United States?**

In [6]:
airports['tzone'].unique()

array(['America/New_York', 'America/Chicago', 'America/Los_Angeles',
       'America/Vancouver', 'America/Phoenix', 'America/Anchorage',
       'America/Denver', 'Pacific/Honolulu', 'Asia/Chongqing', '\\N'],
      dtype=object)

As you can see from the unique values on 'tzone' (time zone).

There are none America cities. We will exclude Asia in a new dataframe.

In [7]:
ap = airports[~airports.tzone.str.contains('Asia')]

In [8]:
ap.sort_values('lat', ascending=False).head()

faa                                        name        lat         lon  \
417  EEN                     Dillant Hopkins Airport  72.270833   42.898333   
230  BRW                  Wiley Post Will Rogers Mem  71.285446 -156.766003   
110  AIN                          Wainwright Airport  70.638056 -159.994722   
708  K03                               Wainwright As  70.613378 -159.860350   
152  ATK  Atqasuk Edward Burnell Sr Memorial Airport  70.467300 -157.436000   

     alt   tz dst              tzone  
417  149 -5.0   A                 \N  
230   44 -9.0   A  America/Anchorage  
110   41 -9.0   A  America/Anchorage  
708   35 -9.0   A  America/Anchorage  
152   96 -9.0   A  America/Anchorage

As you can see from the sorted list above, the two northernmost airports by latitude (the 'lat' column) are:
1. Dillant Hopkins Airport
2. Wiley Post Will Rogers Memorial Airport

A quick search on Google maps returns the following results.
Wiley Post Will Rogers Memorial Airport is the northern most airport that belongs to the US.

**However, Dillant Hopkins Airport is the northernmost airport in the continental US.**

![Wiley](http://pichost.org/images/2018/09/30/wiley_ap.png)
![Dillant](http://pichost.org/images/2018/09/30/dillant_ap.png)

## **2. What is the easternmost airport in the United States?**

We will use the the dataframe where we exluded Asia.

In [9]:
ap.sort_values('lon', ascending=False).head()

faa                        name        lat         lon  alt   tz dst  \
1290  SYA                Eareckson As  52.712275  174.113620   98 -9.0   A   
417   EEN     Dillant Hopkins Airport  72.270833   42.898333  149 -5.0   A   
444   EPM  Eastport Municipal Airport  44.910111  -67.012694   45 -5.0   A   
624   HUL                Houlton Intl  46.123083  -67.792056  489 -5.0   A   
259   CAR                Caribou Muni  46.871500  -68.017917  626 -5.0   A   

                  tzone  
1290  America/Anchorage  
417                  \N  
444    America/New_York  
624    America/New_York  
259    America/New_York

We will sort by longitude this time.

From the new list we can see that Eareckson and Dillant Hopkins Airport are the top two results.

While Eareckson is the easternmost from the prime meridian based on longitude. It is not the easternmost airport in the US.

**Dillant Hopkins Airport is the easternmost airport in the continental US.**

![Dillant](http://pichost.org/images/2018/09/30/dillant_ap.png)
![Earekson](http://pichost.org/images/2018/09/30/ear_ap.png)

## **3. On February 12th, 2013, which New York area airport had the windiest weather?**

Lets revisit the weather dataframe.

In [10]:
weather.head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  wind_speed  \
0    EWR  2013      1    1     1  39.02  26.06  59.37     270.0    10.35702   
1    EWR  2013      1    1     2  39.02  26.96  61.63     250.0     8.05546   
2    EWR  2013      1    1     3  39.02  28.04  64.43     240.0    11.50780   
3    EWR  2013      1    1     4  39.92  28.04  62.21     250.0    12.65858   
4    EWR  2013      1    1     5  39.02  28.04  64.43     260.0    12.65858   

   wind_gust  precip  pressure  visib             time_hour  
0        NaN     0.0    1012.0   10.0  2013-01-01T06:00:00Z  
1        NaN     0.0    1012.3   10.0  2013-01-01T07:00:00Z  
2        NaN     0.0    1012.5   10.0  2013-01-01T08:00:00Z  
3        NaN     0.0    1012.2   10.0  2013-01-01T09:00:00Z  
4        NaN     0.0    1011.9   10.0  2013-01-01T10:00:00Z

In [11]:
weather.tail()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
26110    LGA  2013     12   30    14  35.96  19.94  51.78     340.0   
26111    LGA  2013     12   30    15  33.98  17.06  49.51     330.0   
26112    LGA  2013     12   30    16  32.00  15.08  49.19     340.0   
26113    LGA  2013     12   30    17  30.92  12.92  46.74     320.0   
26114    LGA  2013     12   30    18  28.94  10.94  46.41     330.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
26110    13.80936   21.86482     0.0    1017.1   10.0  2013-12-30T19:00:00Z  
26111    17.26170   21.86482     0.0    1018.8   10.0  2013-12-30T20:00:00Z  
26112    14.96014   23.01560     0.0    1019.5   10.0  2013-12-30T21:00:00Z  
26113    17.26170        NaN     0.0    1019.9   10.0  2013-12-30T22:00:00Z  
26114    18.41248        NaN     0.0    1020.9   10.0  2013-12-30T23:00:00Z

**What is the difference between wind speed and wind speed?**

According to [www.differencebetween.net]('http://www.differencebetween.net/science/nature/difference-between-gust-and-wind/')

`"All gusts are a type of wind. A gust is a sudden increase of the wind's speed that lasts no more than 20 seconds. This usually occurs when wind speeds reach a peak of at least 16 knots. A wind gust usually comes in 2-minute intervals."`

We will not be using wind_gust.

In [12]:
weather['origin'].unique()

array(['EWR', 'JFK', 'LGA'], dtype=object)

Here we have three major airports in the New York area:

1. EWR (Newark Liberty International Airport)
2. JFK (John F. Kennedy International Airport)
3. LGA (LaGuardia Airport)

In [13]:
weather.isnull().sum()

origin            0
year              0
month             0
day               0
hour              0
temp              1
dewp              1
humid             1
wind_dir        460
wind_speed        4
wind_gust     20778
precip            0
pressure       2729
visib             0
time_hour         0
dtype: int64

The output shows the overall NULL (or NaN) values on all columns.

We will focus on wind_speed since wind_gust is mostly NULL and it does not affect analysis.

Lets sort the dataframe in descening order on 'wind_speed', from highgest to lowest.

In [14]:
weather.sort_values('wind_speed', ascending=False).head()

origin  year  month  day  hour   temp   dewp   humid  wind_dir  \
1009     EWR  2013      2   12     3  39.02  26.96   61.63     260.0   
9425     JFK  2013      1   31     4  53.60  53.06  100.00     200.0   
724      EWR  2013      1   31     6  57.20  53.60   87.74     270.0   
722      EWR  2013      1   31     4  60.80  59.00   93.79     230.0   
18131    LGA  2013      1   31     4  59.00  55.40   93.69     230.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
1009   1048.36058        NaN    0.00    1008.3   10.0  2013-02-12T08:00:00Z  
9425     42.57886   52.93588    0.20       NaN    2.0  2013-01-31T09:00:00Z  
724      42.57886   55.23744    0.03       NaN   10.0  2013-01-31T11:00:00Z  
722      40.27730   58.68978    0.30       NaN    6.0  2013-01-31T09:00:00Z  
18131    40.27730   55.23744    0.41       NaN    8.0  2013-01-31T09:00:00Z

Let's be sure and isolate the date of February 12, 2013.

In [15]:
we = weather[weather.time_hour.str.contains('2013-02-12')]

In [16]:
we.sort_values('wind_speed', ascending=False).head()

origin  year  month  day  hour   temp   dewp  humid  wind_dir  \
1009     EWR  2013      2   12     3  39.02  26.96  61.63     260.0   
18417    LGA  2013      2   12     2  42.98  26.06  50.94     290.0   
1018     EWR  2013      2   12    12  44.06  26.06  48.87     270.0   
18428    LGA  2013      2   12    13  44.06  23.00  43.02     300.0   
9717     JFK  2013      2   12     8  42.08  24.08  48.56     280.0   

       wind_speed  wind_gust  precip  pressure  visib             time_hour  
1009   1048.36058        NaN     0.0    1008.3   10.0  2013-02-12T08:00:00Z  
18417    23.01560   31.07106     0.0    1007.1   10.0  2013-02-12T07:00:00Z  
1018     21.86482   31.07106     0.0    1012.5   10.0  2013-02-12T17:00:00Z  
18428    21.86482   25.31716     0.0    1011.7   10.0  2013-02-12T18:00:00Z  
9717     20.71404   27.61872     0.0    1011.6   10.0  2013-02-12T13:00:00Z

As you can see on 2013-02-12 at hour 3, Newark had the windiest weather in the NY area.

[It is historically recorded in www.wunderground.com](https://www.wunderground.com/history/daily/us/nj/newark/KEWR/date/2013-2-12)

![Historic_Data](http://pichost.org/images/2018/09/30/hist.png)

As you can see in 3:51 AM the wind speed is recorded as 1048 mph!

In [17]:
we.groupby('origin').wind_speed.max()

origin
EWR    1048.36058
JFK      20.71404
LGA      23.01560
Name: wind_speed, dtype: float64

Max wind speed from all three airports.

## Findings:

1. What is the northernmost airport in the United States? Dillant Hopkins Airport

    Dillant Hopkins Airport

2. What is the easternmost airport in the United States? 
    
    Dillant Hopkins Airport

3. On February 12th, 2013, which New York area airport had the windiest weather?
   
   Newark Airport (EWR) had a wind speed of 1048mph on February 2, 2013.